### PR TITLE
[receiver/prometheusreceiver] implement append native histogram

### LINKED ISSUE #26555: Prometheus receiver is not enabling native histograms feature
### Component(s)

receiver/prometheus

## Description
Prometheus receiver is not enabling native histograms feature. We have a usecase where we are using Prometheus receiver and PRW exporter to export metrics. 

## Steps to Reproduce
I have verified that Prometheus receiver is not able to scrap native histogram metrics while vanilla prometheus works with enabling [native histogram feature](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms).

## Expected Result
Native histogram should be able to be scraped.

## Actual Result
Not able to see native histogram after verifying from logging exporter.

### Collector version

0.80

### PR DESCRIPTION
**Pull request by krajorama**
**Description:** 

Implement native histogram append MVP.
Very similar to appending a float sample.

Limitations:
- Only support integer counter histograms fully.
- In case a histogram has both classic and native buckets, we only store one of them. Governed by scrape_classic_histograms scrape option. The reason is that in the OTEL model the metric family is identified by the normalized name (without _count, _sum, _bucket suffixes for the classic histograms), meaning that the classic and native histograms would map to the same metric family in OTEL model , but that cannot have both Histogram and ExponentialHistogram types at the same time.
- Gauge histograms are dropped with warning as that temporality is unsupported, see https://github.com/open-telemetry/opentelemetry-specification/issues/2714
- NoRecordedValue attribute might be unreliable. Prometheus scrape marks all series with float NaN values when stale, but transactions in prometheusreceiver are stateless, meaning that we have to use heuristics to figure out if we need to add a NoRecordedValue data point to an Exponential Histogram metric. (Need work in Prometheus.)

Additionally: 
- Created timestamp supported.
- Float counter histograms not fully tested and lose precision, but we don't expect instrumentation to expose these anyway.

**Link to tracking Issue:**

Fixes: #26555 

**Testing:** 

Added unit tests and e2e tests.

**Documentation:**

TBD: will have to call out protobuf negotiation while no text format. #27030

### CODE REVIEW

File: receiver/prometheusreceiver/internal/transaction.go [hunk 1]
```diff
@@ -203,9 +203,63 @@ func (t *transaction) AppendExemplar(_ storage.SeriesRef, l labels.Labels, e exe
...
+
+	// See https://www.prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series
+	// up: 1 if the instance is healthy, i.e. reachable, or 0 if the scrape failed.
+	// But it can also be a staleNaN, which is inserted when the target goes away.
+	// Native histogram: the up metric is generated as a counter so we can assume that
+	// this metric is user defined. Do explicit nothing with up metric to be consistent with Append()
+	// if metricName == scrapeUpMetricName {}
+
+	// For the `target_info` metric we need to convert it to resource attributes.
+	if metricName == targetMetricName {
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/transaction.go:245]: target_info also won't ever be a histogram

File: receiver/prometheusreceiver/internal/transaction.go [hunk 2]
```diff
@@ -203,9 +203,63 @@ func (t *transaction) AppendExemplar(_ storage.SeriesRef, l labels.Labels, e exe
...
+		t.AddTargetInfo(ls)
+		return 0, nil
+	}
+
+	// For the `otel_scope_info` metric we need to convert it to scope attributes.
+	if metricName == scopeMetricName {
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/transaction.go:251]: scope info also won't ever be a histogram.

**Comment by krajorama** [comment]
Seems like I need to create a new [metricFamily](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/55afb0a9fdcbc0d7f2e051302d38dafc384c3505/receiver/prometheusreceiver/internal/metricfamily.go#L29) that has type [MetricTypeExponentialHistogram](https://github.com/open-telemetry/opentelemetry-collector/blob/6405e152a2d9a08de1fe1b70fd4ac775a12a3cda/pdata/pmetric/metric_type.go#L15) .

The only issue seems like that Prometheus and Grafana Agent supports scraping classic and native (exponential) histogram version of the same metric, making the key (which is the normalized metric name) the same for the classic and native histogram. So I'm not totally sure what to do here. Maybe reuse the metric family, but make it possible to have a [metricGroup](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/55afb0a9fdcbc0d7f2e051302d38dafc384c3505/receiver/prometheusreceiver/internal/metricfamily.go#L42) inside that has the exponential histogram. This would be possible to route for prometheus remote write exporter, but I doubt it would work for otel.

**Comment by dashpole** [comment]
> do not enable scraping classic histograms when there is native histogram. Will not cause an issue for OTEL. Only people that would get bit by this is someone trying to switch from prometheus to otel but using classic + native histograms at the same time (i.e. migrating).

Could we add config to control whether we keep the native vs the fixed bucket histogram?

Is there any existing configuration in the prometheus server that controls whether native histograms are enabled or not?

**Comment by dashpole** [comment]
Ideally we would only scrape native histograms if the feature flag is enabled, and only scrape classic histograms if scrape_classic_histograms is set.

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 3]
```diff
@@ -156,6 +159,95 @@ func (mg *metricGroup) toDistributionPoint(dest pmetric.HistogramDataPointSlice)
...
+	case mg.fhValue != nil:
+		fh := mg.fhValue
+		// Input is a float native histogram. This conversion will lose
+		// precision,but we don't actually expect float histograms in scrape,
+		// since these are typically the result of operations on integer
+		// native histograms in the database.
+		point.SetScale(fh.Schema)
+		point.SetCount(uint64(fh.Count))
+		point.SetSum(fh.Sum)
+		point.SetZeroCount(uint64(fh.ZeroCount))
```
Comment (krajorama) [inline receiver/prometheusreceiver/internal/metricfamily.go:182]: There is no point.SetZeroThreshold function :(

**Comment by krajorama** [comment]
@dashpole I've opened another PR #29153 to set up for testing native histogram scrape here e2e, ptal.

**Comment by krajorama** [comment]
Rebased onto #29153 and added first e2e test.

**Comment by krajorama** [comment]
> IIUC, the sum should not be reported if it includes negative observations (at least for "standard" histograms, since sum is a counter). https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#histogram. Is sum not required to be monotonic for native histograms?

I made a simple test with observing some negative values and checked /metrics:
```
# HELP golang_manual_histogram This is a histogram with manually selected parameters
# TYPE golang_manual_histogram histogram
golang_manual_histogram_bucket{address="0.0.0.0",generation="20",port="5001",le="-5"} 1
golang_manual_histogram_bucket{address="0.0.0.0",generation="20",port="5001",le="-1"} 2
golang_manual_histogram_bucket{address="0.0.0.0",generation="20",port="5001",le="1"} 3
golang_manual_histogram_bucket{address="0.0.0.0",generation="20",port="5001",le="5"} 3
golang_manual_histogram_bucket{address="0.0.0.0",generation="20",port="5001",le="+Inf"} 3
golang_manual_histogram_sum{address="0.0.0.0",generation="20",port="5001"} -12
golang_manual_histogram_count{address="0.0.0.0",generation="20",port="5001"} 3
```
To quote the [documentation](https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations) : "The sum of observations (showing up as a time series with a _sum suffix) behaves like a counter, too, as long as there are no negative observations. "

So there is no guarantee that the sum actually behaves like a counter.


> 
> > On the other hand it is possible that the Count increases, but there's actually a reduction in individual buckets, so in theory the code should be looking at bucket counts.
> 
> Can you clarify? Are you talking about buckets being merged? Or about the histogram resetting, but receiving more observations than it previously had since the reset?

The second case. Let's suppose you have 4 observations in the 0th offset positive bucket and that's the only bucket in use, so your count is 4. For the next scrape , schema doesn't change , and you get 2 observations in the 0th, 1st, 2nd positive buckets, so your overall count is 6, but obviously there was a restart since 0th bucket count went down.

See code starting from https://github.com/prometheus/prometheus/blob/1bfb3ed062e99bd3c74e05d9ff9a7fa4e30bbe21/tsdb/chunkenc/histogram.go#L272  . 

But this is all academic if it's not actually that important to detect these cases :)

**Comment by dashpole** [comment]
> The second case. Let's suppose you have 4 observations in the 0th offset positive bucket and that's the only bucket in use, so your count is 4. For the next scrape , schema doesn't change , and you get 2 observations in the 0th, 1st, 2nd positive buckets, so your overall count is 6, but obviously there was a restart since 0th bucket count went down.

Those cases seem rare in practice, and we've accepted those thus far as unsolvable. I'm OK with not properly handling those.

It sounds like using count to detect resets is probably our best bet.

IIRC, prom protobuf is going to begin include start timestamps soon, so hopefully these problems are obsoleted over time.

**Comment by krajorama** [comment]
> > The second case. Let's suppose you have 4 observations in the 0th offset positive bucket and that's the only bucket in use, so your count is 4. For the next scrape , schema doesn't change , and you get 2 observations in the 0th, 1st, 2nd positive buckets, so your overall count is 6, but obviously there was a restart since 0th bucket count went down.
> 
> Those cases seem rare in practice, and we've accepted those thus far as unsolvable. I'm OK with not properly handling those.
> 
> It sounds like using count to detect resets is probably our best bet.
> 
> IIRC, prom protobuf is going to begin include start timestamps soon, so hopefully these problems are obsoleted over time.

Understood, the one thing we have to make sure is that the exported/prometheusremotewrite MUST NOT set the the [ResetHint](https://github.com/prometheus/prometheus/blob/cf528bef034dcf078d3bff0d50224a13068c391e/prompb/types.proto#L64) to "NO", unless it's absolutely sure that there was no reset (as defined by Prometheus). [Currently](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9de22efdd7ffef7c2de3a4ac2613d874ffcb7f3a/pkg/translator/prometheusremotewrite/histograms.go#L76) the hint is left to "UNKOWN" which is fine.

So basically it would be possible to set "NO" only with the create timestamp, since the current algorithm above is not compatible with the Prometheus definition of counter reset. cc @Aneurysm9 (as maintainer of the exporter).

**Comment by krajorama** [comment]
@dashpole this is mostly finished, I want to add a reference to a Prometheus issue we need to create and also test in real life with a couple of thousands of native histogram metrics.

**Comment by dashpole** [comment]
> Understood, the one thing we have to make sure is that the exported/prometheusremotewrite MUST NOT set the the [ResetHint](https://github.com/prometheus/prometheus/blob/cf528bef034dcf078d3bff0d50224a13068c391e/prompb/types.proto#L64) to "NO", unless it's absolutely sure that there was no reset (as defined by Prometheus). [Currently](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9de22efdd7ffef7c2de3a4ac2613d874ffcb7f3a/pkg/translator/prometheusremotewrite/histograms.go#L76) the hint is left to "UNKOWN" which is fine.

> So basically it would be possible to set "NO" only with the create timestamp, since the current algorithm above is not compatible with the Prometheus definition of counter reset.

Would you mind adding a comment in the exporter(s), referencing prometheus' definition of reset to ensure that isn't changed?  We could explicitly set it to unknown to make it clear that it is the correct value for us.  It can be done separately from this PR.

**Comment by krajorama** [comment]
> Would you mind adding a comment in the exporter(s), referencing prometheus' definition of reset to ensure that isn't changed? We could explicitly set it to unknown to make it clear that it is the correct value for us. It can be done separately from this PR.

Hi, created PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29326

Also this PR is now ready for review.

File: receiver/prometheusreceiver/internal/transaction.go [hunk 4]
```diff
@@ -143,16 +143,34 @@ func (t *transaction) Append(_ storage.SeriesRef, ls labels.Labels, atMs int64,
...
+		} else {
+			t.logger.Warn("failed to add datapoint", zap.Error(err), zap.String("metric_name", metricName), zap.Any("labels", ls))
+		}
 	}
 
 	return 0, nil // never return errors, as that fails the whole scrape
 }
 
-func (t *transaction) getOrCreateMetricFamily(scope scopeID, mn string) *metricFamily {
+func (t *transaction) getOrCreateMetricFamily(scope scopeID, mn string) (*metricFamily, bool) {
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/transaction.go:173]:
```suggestion
// getOrCreateMetricFamily returns the metric family for the given metric name and scope,
// and true if an existing family was found.
func (t *transaction) getOrCreateMetricFamily(scope scopeID, mn string) (*metricFamily, bool) {
```

File: receiver/prometheusreceiver/internal/transaction.go [hunk 5]
```diff
@@ -197,15 +216,64 @@ func (t *transaction) AppendExemplar(_ storage.SeriesRef, l labels.Labels, e exe
...
+	}
+
+	// The `up`, `target_info`, `otel_scope_info` metrics should never generate native histograms,
+	// thus we don't check for them here as opposed to the Append function.
+
+	curMF, existing := t.getOrCreateMetricFamily(getScopeID(ls), metricName)
+	if !existing {
+		curMF.mtype = pmetric.MetricTypeExponentialHistogram
+	} else if curMF.mtype != pmetric.MetricTypeExponentialHistogram {
+		// Already scraped as classic histogram.
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/transaction.go:263]: is the classic histogram append guaranteed to come first?

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 6]
```diff
@@ -156,6 +159,115 @@ func (mg *metricGroup) toDistributionPoint(dest pmetric.HistogramDataPointSlice)
...
+	} else {
+		// metrics_adjuster adjusts the startTimestamp to the initial scrape timestamp
+		point.SetStartTimestamp(tsNanos)
+	}
+	point.SetTimestamp(tsNanos)
+	populateAttributes(pmetric.MetricTypeHistogram, mg.ls, point.Attributes())
+	mg.setExemplars(point.Exemplars())
+}
+
+func convertDeltaBuckets(spans []histogram.Span, deltas []int64, buckets pcommon.UInt64Slice) {
```
[the remark this answers is not in the document]
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metricfamily.go:237]: Depends on the proto, which can carry either integer counts or float counts. The integer counts are delta encoded, while floats are absolute. Instrumentation should carry integer counts, float counts are currently created when series are operated on in Prometheus in recording rules. But nothing is stopping the target from exposing float counts in theory.

File: receiver/prometheusreceiver/README.md [hunk 7]
```diff
@@ -111,6 +111,7 @@ The prometheus receiver also supports additional top-level options:
...
 - **use_start_time_metric**: When set to true, this enables retrieving the start time of all counter metrics from the process_start_time_seconds metric. This is only correct if all counters on that endpoint started after the process start time, and the process is the only actor exporting the metric after the process started. It should not be used in "exporters" which export counters that may have started before the process itself. Use only if you know what you are doing, as this may result in incorrect rate calculations. Defaults to false.
 - **start_time_metric_regex**: The regular expression for the start time metric, and is only applied when use_start_time_metric is enabled.  Defaults to process_start_time_seconds.
+- **enable_protobuf_negotiation**: When set to true, Prometheus receiver will try to negotiate the protobuf format first for scraping the target. This is considered experimental.
```
[the remark this answers is not in the document]
Comment (krajorama) [reply receiver/prometheusreceiver/README.md:114]: Prometheus will enable two things by default eventually (I'd guess next year):
- try Protobuf negotiation first (since it's more efficient than text)
- allow native histograms (currently feature flag)

Actually the text format for native histograms is being worked on currently, so again eventually it will not matter what you set protobuf negotiation to.

Given the current code, to follow Prometheus, the default for `enable_protobuf_negotiation` option should be switched to `true` in otel collector as well when Prometheus makes the change.

The real "problem" will be native histograms being on by default. Since for a metric that has both classic and native buckets, we can only scrape one due to the name conflict (both result in the same metric name). So if we start enabling native histograms by default, suddenly the classic buckets would be gone, replaced by exponential buckets.

[the remark this answers is not in the document]
Comment (dashpole) [reply receiver/prometheusreceiver/README.md:114]: So, we would want our eventual configuration to only include a knob for switching between fixed-bucket and exponential histogram, right?  Do we need to keep config for protobuf negotiation around long-term?  It sounds like we essentially just want to keep the default histogram format as fixed-bucket, rather than exponential

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 8]
```diff
@@ -311,6 +423,25 @@ func (mf *metricFamily) addSeries(seriesRef uint64, metricName string, ls labels
...
+	if mg.mtype != pmetric.MetricTypeExponentialHistogram {
+		return fmt.Errorf("metric type mismatch for exponential histogram metric %v type %s", metricName, mg.mtype.String())
+	}
+	if h != nil {
+		mg.count = float64(h.Count)
+		mg.sum = h.Sum
+		mg.hasCount = true
+		mg.hasSum = true
+		mg.hValue = h
+		mg.fhValue = fh
```
[the remark this answers is not in the document]
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metricfamily.go]: nice catch, fixed. that's what I get for not writing a unit test for it...

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 9]
```diff
@@ -156,6 +159,115 @@ func (mg *metricGroup) toDistributionPoint(dest pmetric.HistogramDataPointSlice)
...
+		return
+	}
+	point := dest.AppendEmpty()
+	point.SetTimestamp(timestampFromMs(mg.ts))
+
+	// We do not set Min or Max as native histograms don't have that information.
+	switch {
+	case mg.fhValue != nil:
+		fh := mg.fhValue
+		point.SetScale(fh.Schema)
```
[the remark this answers is not in the document]
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metricfamily.go]: copy paste error, fixed

File: receiver/prometheusreceiver/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (krajorama) [reply receiver/prometheusreceiver/README.md:114]: > So, we would want our eventual configuration to only include a knob for switching between fixed-bucket and exponential histogram, right? Do we need to keep config for protobuf negotiation around long-term? It sounds like we essentially just want to keep the default histogram format as fixed-bucket, rather than exponential

I guess so. The question that comes up for Prometheus is different, because you can have the classic histograms beside the native histograms (as names are different). Still the issue is similar in that if we enable native histograms by default, you suddenly have a breaking change by switching from classic buckets to native buckets. And maybe we don't need an extra setting since you can set `scrape_classic_histograms` in the scrape config. In Prometheus that gets you the classic buckets **as well** and in otel collector it would get you the classic buckets **only**. At least in the my current implementation.

The whole thing only affects cases where the metric has both classic and native defined.

cc @beorn7

File: receiver/prometheusreceiver/internal/transaction.go (hunk 5 above)
Comment (krajorama) [reply receiver/prometheusreceiver/internal/transaction.go:265]: It's the other way around. No guarantee beyond that this is how it's [coded](https://github.com/prometheus/prometheus/blob/b37258c99b901490ccf22f85f472ba2ba8c8be33/model/textparse/protobufparse.go#L72) right now, which is why I added the test case `TestNativeVsClassicHistogramScrapeViaProtobuf` .

**Comment by krajorama** [comment]
I've talked to the potential tester of this PR and they said earliest February 2024. On the other hand the use case will be to scrape Grafana Mimir with it, so when I have time I can do the testing myself or someone from our team. Hope to get some movement on this soon.

**Comment by jmcarp** [comment]
This is exciting! If this gets merged, would the prometheusremotewrite exporter be able to export native histograms, or would that require additional code changes?

**Comment by jmcarp** [comment]
Actually, @krajorama what kind of testing are you looking for? I'm interested in using this feature to scrape native histograms from an application, then forward to mimir as native histograms. I might have time to test this out if it would be helpful.

**Comment by krajorama** [comment]
> Actually, @krajorama what kind of testing are you looking for? I'm interested in using this feature to scrape native histograms from an application, then forward to mimir as native histograms. I might have time to test this out if it would be helpful.

Hi @jmcarp, sorry about the long wait, was/is busy with other stuff. The use case is to scrape Mimir itself as it has a native histogram exposed `cortex_request_duration_seconds` (I'm working on adding more actually.) But any other application would be good as well that has native histograms, the point is to just go beyond unit tests and run it for some time and try restarts, realistic load - basically give it a spin. 

The bad news is that it badly needs a rebase/merge from main now.

**Comment by krajorama** [comment]
> This is exciting! If this gets merged, would the prometheusremotewrite exporter be able to export native histograms, or would that require additional code changes?

I have used the prometheusremotewrite exporter to write exponentil histograms to Prometheus/Mimir so that should indeed work.

**Comment by dashpole** [comment]
If testing will take more time, we can always merge it behind an alpha feature-gate to give time for testing.  That way you don't need to keep rebasing it.

**Comment by jmcarp** [comment]
> we can always merge it behind an alpha feature-gate to give time for testing. That way you don't need to keep rebasing it.

Sounds great to me!

**Comment by krajorama** [comment]
To my surprise the rebase went smoothly. I'm going to give it a smoke test and try adding feature-gate

**Comment by krajorama** [comment]
I've done some testing and found that I can now set ZeroThreshold in the otel model so that's now implemented.
Also found a bug of not setting created timestamp correctly from _created metric.

Also ran into an issue with exemplars:
```
2024-03-16T20:03:44.340+0100	error	exporterhelper/queue_sender.go:97	Exporting failed. Dropping data.	{"kind": "exporter", "data_type": "metrics", "name": "prometheusremotewrite", "error": "Permanent error: Permanent error: Permanent error: remote write returned HTTP status 400 Bad Request; err = %!w(<nil>): failed pushing to ingester 5444ebe325cd: user=anonymous: err: out of order exemplar. timestamp=2024-03-16T19:02:51.834Z, series=cortex_request_duration_seconds{container=\"mimir-2\", instance=\"mimir-2:8001\", job=\"mimir-2\", method=\"gRPC\", route=\"/cortex.Inges", "dropped_items": 883}
```

This is similar to https://github.com/prometheus/prometheus/pull/13021 where the fix was to ensure that exemplars have a timestamp (afaik otel already ensures this) and timestamps are in ascending order.

**Comment by krajorama** [comment]
> Also ran into an issue with exemplars:
> 
> ```
> 2024-03-16T20:03:44.340+0100	error	exporterhelper/queue_sender.go:97	Exporting failed.
> Dropping data.	{"kind": "exporter", "data_type": "metrics", "name": "prometheusremotewrite",
> "error": "Permanent error: Permanent error: Permanent error: remote write returned HTTP status 400 Bad Request;
> err = %!w(<nil>): failed pushing to ingester 5444ebe325cd: user=anonymous: err: out of order exemplar.
> timestamp=2024-03-16T19:02:51.834Z, series=cortex_request_duration_seconds{container=\"mimir-2\",
> instance=\"mimir-2:8001\", job=\"mimir-2\", method=\"gRPC\", route=\"/cortex.Inges", "dropped_items": 883}
> ```
> 
> This is similar to [prometheus/prometheus#13021](https://github.com/prometheus/prometheus/pull/13021) where the fix was to ensure that exemplars have a timestamp (afaik otel already ensures this) and timestamps are in ascending order.

Actually this might be a Mimir side bug: https://github.com/grafana/mimir/pull/7640

File: receiver/prometheusreceiver/metrics_receiver_protobuf_test.go [hunk 10]
```diff
@@ -156,3 +158,447 @@ func TestScrapeViaProtobuf(t *testing.T) {
...
+					compareMetricUnit(""),
+					[]dataPointExpectation{{
+						exponentialHistogramComparator: []exponentialHistogramComparator{
+							compareExponentialHistogram(1214, 3456, 5, -3, []uint64{1, 0, 2}, 2, []uint64{1, 0, 0, 1}),
+						},
+					}},
+				),
+			},
+		},
+		"feature disable scrape classic on": {
```
Comment (krajorama) [inline receiver/prometheusreceiver/metrics_receiver_protobuf_test.go:360]:
```suggestion
		"feature disabled scrape classic on": {
```

**Comment by krajorama** [comment]
> If testing will take more time, we can always merge it behind an alpha feature-gate to give time for testing. That way you don't need to keep rebasing it.

@dashpole I've added a feature-gate to this, but now you can get into a weird situation where you
1) turn on protobuf negotiation
2) keep feature gate disabled
3) don't have `scrape_classic_histograms` enabled in Prometheus scrape config
then
1) if metric has only classic buckets, it works fine
2) if metric has only native buckets, it is not stored - as feature is off
3) BUT if metric has both classic and native buckets, it's not scraped as classic and not stored as native , so it's kind of lost

I wonder if I should refuse this configuration with an error maybe?

File: receiver/prometheusreceiver/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (dashpole) [reply receiver/prometheusreceiver/README.md:115]: Enabling protobuf negotiation can now be done using the prometheus scrape config.  See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30934.

**Comment by dashpole** [comment]
@krajorama you could just make the feature gate take precedence over scrape_classic_histogram.  If the feature gate is disabled, it should always keep the current behavior, even when scrape_classic_histogram is false.

**Comment by krajorama** [comment]
> @krajorama you could just make the feature gate take precedence over scrape_classic_histogram. If the feature gate is disabled, it should always keep the current behavior, even when scrape_classic_histogram is false.

Ack, that shouldn't be hard to implement.

**Comment by krajorama** [comment]
OTOH To be able to efficiently validate exemplars in Mimir (https://github.com/grafana/mimir/pull/7640 ) they would have to arrive in order. But that's not guaranteed by the otel collector . I'm testing what happens if we sort the exemplars on ingest - but I'm worried of introducing a lot of overhead

**Comment by krajorama** [comment]
> > @krajorama you could just make the feature gate take precedence over scrape_classic_histogram. If the feature gate is disabled, it should always keep the current behavior, even when scrape_classic_histogram is false.
> 
> Ack, that shouldn't be hard to implement.

I've added code to enforce scraping classic histograms if native histograms feature is off. (Note: we'll have to do the same in Prometheus when text format of native histograms becomes a thing.)

PR ready for review , pinging people that commented on it: @dashpole @fatsheep9146 @Aneurysm9 @jpkrohling 

By the way, Prometheus 0.49 had some fixes regarding scrape, so I'll make a separate PR to update that dependency. Most notably https://github.com/prometheus/prometheus/pull/13021 .

File: receiver/prometheusreceiver/internal/metrics_adjuster.go [hunk 11]
```diff
@@ -346,6 +355,54 @@ func (a *initialPointAdjuster) adjustMetricHistogram(tsm *timeseriesMap, current
...
+		if !found {
+			// initialize everything.
+			tsi.histogram.startTime = currentDist.StartTimestamp()
+			tsi.histogram.previousCount = currentDist.Count()
+			tsi.histogram.previousSum = currentDist.Sum()
+			continue
+		}
+
+		if currentDist.Flags().NoRecordedValue() {
+			// TODO: Investigate why this does not reset.
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metrics_adjuster.go:386]: Is this still relevant?

File: receiver/prometheusreceiver/internal/transaction.go [hunk 12]
```diff
@@ -143,16 +146,36 @@ func (t *transaction) Append(_ storage.SeriesRef, ls labels.Labels, atMs int64,
...
+
+	if t.enableNativeHistograms && curMF.mtype == pmetric.MetricTypeExponentialHistogram {
+		curMF.mtype = pmetric.MetricTypeHistogram
+	}
+
+	seriesRef := t.getSeriesRef(ls, curMF.mtype)
+	err := curMF.addSeries(seriesRef, metricName, ls, atMs, val)
 	if err != nil {
-		t.logger.Warn("failed to add datapoint", zap.Error(err), zap.String("metric_name", metricName), zap.Any("labels", ls))
+		// Handle special case of float sample indicating staleness of native
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/transaction.go:158]: trying to wrap my head around this.  Why does the empty label error indicate is is a stale exponential histogram?

File: receiver/prometheusreceiver/internal/transaction.go [hunk 13]
```diff
@@ -143,16 +146,36 @@ func (t *transaction) Append(_ storage.SeriesRef, ls labels.Labels, atMs int64,
...
+		// histogram. This is similar to how Prometheus handles it, but we
+		// don't have access to the previous value so we're applying some
+		// heuristics to figure out if this is native histogram or not.
+		if t.enableNativeHistograms && errors.Is(err, errEmptyLeLabel) && !existing && value.IsStaleNaN(val) && curMF.mtype == pmetric.MetricTypeHistogram {
+			mg := curMF.loadMetricGroupOrCreate(seriesRef, ls, atMs)
+			curMF.mtype = pmetric.MetricTypeExponentialHistogram
+			mg.mtype = pmetric.MetricTypeExponentialHistogram
+			_ = curMF.addExponentialHistogramSeries(seriesRef, metricName, ls, atMs, &histogram.Histogram{Sum: val}, nil)
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/transaction.go:166]: nit: it would be less confusing to just use the staleNaN directly here.

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 14]
```diff
@@ -311,6 +430,33 @@ func (mf *metricFamily) addSeries(seriesRef uint64, metricName string, ls labels
...
+		mg.hasSum = true
+		mg.hValue = nil
+		mg.fhValue = fh
+	case h != nil:
+		mg.count = float64(h.Count)
+		mg.sum = h.Sum
+		mg.hasCount = true
+		mg.hasSum = true
+		mg.hValue = h
+		mg.fhValue = nil
```
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metricfamily.go:455]: is setting the value to nil here neccessary?

File: .chloggen/prometheusreceiver-append-native-histogram.yaml [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Allows receiving prometheus native histograms
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [26555]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Native histograms are compatible with otel exponential histograms.
```
Comment (dashpole) [inline .chloggen/prometheusreceiver-append-native-histogram.yaml:18]: You should mention the steps to enable it here

**Comment by krajorama** [comment]
Makes sense to wait for #30934 to finalize documentation side and get bugfix https://github.com/prometheus/prometheus/pull/13021 as well.

File: receiver/prometheusreceiver/internal/metricfamily.go (hunk 14 above)
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metricfamily.go:455]: It's defensive programing. It should not happen that we scrape something in two ways. I'll convert to an error instead if a different precision histogram is already set to be more clear about this.

File: receiver/prometheusreceiver/internal/metrics_adjuster.go (hunk 11 above)
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metrics_adjuster.go:386]: I don't know, it's copy paste from `adjustMetricHistogram`. Since I didn't quite understand it and I pass my tests, I though it's best to just keep it.

File: receiver/prometheusreceiver/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (krajorama) [reply receiver/prometheusreceiver/README.md:121]: Will wait for  PR #30934  and rebase.

File: receiver/prometheusreceiver/internal/transaction.go (hunk 12 above)
Comment (krajorama) [reply receiver/prometheusreceiver/internal/transaction.go:162]: I've added more comment, ptal

File: receiver/prometheusreceiver/internal/metricfamily.go (hunk 14 above)
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metricfamily.go]: Updated with error handling.

File: .chloggen/prometheusreceiver-append-native-histogram.yaml (hunk 15 above)
Comment (krajorama) [reply .chloggen/prometheusreceiver-append-native-histogram.yaml]: done

**Comment by krajorama** [comment]
> Makes sense to wait for #30934 to finalize documentation side and get bugfix [prometheus/prometheus#13021](https://github.com/prometheus/prometheus/pull/13021) as well.

I see it was merged. I updated the branch locally, but running into an issue with the handling of `native_histogram_min_bucket_factor` in the scrape config. Not sure if I need to fix something in Promtheus or here yet.

**Comment by krajorama** [comment]
> > Makes sense to wait for #30934 to finalize documentation side and get bugfix [prometheus/prometheus#13021](https://github.com/prometheus/prometheus/pull/13021) as well.
> 
> I see it was merged. I updated the branch locally, but running into an issue with the handling of `native_histogram_min_bucket_factor` in the scrape config. Not sure if I need to fix something in Promtheus or here yet.

It's a bug in Promtheus. Have to make the fix and get it into otel as well

**Comment by krajorama** [comment]
Depends on https://github.com/prometheus/prometheus/pull/13846 and backport to 0.50.2 at least

**Comment by krajorama** [comment]
> Depends on [prometheus/prometheus#13846](https://github.com/prometheus/prometheus/pull/13846) and backport to 0.50.2 at least

Will be included in 0.51.1 for sure.

File: receiver/prometheusreceiver/internal/metricfamily_test.go [hunk 16]
```diff
@@ -293,6 +294,199 @@ func TestMetricGroupData_toDistributionUnitTest(t *testing.T) {
...
+				attributes := point.Attributes()
+				attributes.PutStr("a", "A")
+				attributes.PutStr("b", "B")
+				return point
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
Comment (fatsheep9146) [inline receiver/prometheusreceiver/internal/metricfamily_test.go:433]: Is this line redundant?

**Comment by krajorama** [comment]
@dashpole Any chance to update Prometheus to [0.51.1](https://github.com/prometheus/prometheus/releases/tag/v2.51.1) ? I don't have the time currently :(

**Comment by dashpole** [comment]
I need to look into https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31572#issuecomment-2029701231

File: receiver/prometheusreceiver/internal/metricfamily_test.go (hunk 16 above)
Comment (krajorama) [reply receiver/prometheusreceiver/internal/metricfamily_test.go:433]: Isn't the loop var fix in Go 1.22 ? We require 1.21 right now. Maybe I'm mixing this up, but there's 3 other tests that have this line (I did copy :) ).

**Comment by krajorama** [comment]
After the Prometheus update tests pass again so I think this is ready for review again. I suggest to focus on whether the feature flag is correctly applied.

File: receiver/prometheusreceiver/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (dashpole) [reply receiver/prometheusreceiver/README.md:115]: I think this can be removed now

File: receiver/prometheusreceiver/internal/transaction.go (hunk 12 above)
Comment (dashpole) [reply receiver/prometheusreceiver/internal/transaction.go:164]: makes sense, thanks.

**Review by dashpole — COMMENTED** [review]
lgtm after enable_protobuf_negotiation is removed from the README

**Review by dashpole — APPROVED** [review]

**Review by fatsheep9146 — APPROVED** [review]

**Comment by krajorama** [comment]
Thank you, I don't have the right to merge. I'll keep updating a branch regularly until merged.
